proxy/proxy.go

```diff
@@ -315,134 +279,57 @@ func tryCatch(p func(), onErr func(err interface{})) {
-func (c *filterContext) PathParam(key string) string         { return c.pathParams[key] }
-func (c *filterContext) StateBag() map[string]interface{}    { return c.stateBag }
-func (c *filterContext) BackendUrl() string                  { return c.backendUrl }
-func (c *filterContext) OriginalRequest() *http.Request      { return c.originalRequest }
-func (c *filterContext) OriginalResponse() *http.Response    { return c.originalResponse }
-func (c *filterContext) OutgoingHost() string                { return c.outgoingHost }
-func (c *filterContext) SetOutgoingHost(h string)            { c.outgoingHost = h }
-
-func (c *filterContext) Serve(res *http.Response) {
-	res.Request = c.Request()
-
-	if res.Header == nil {
-		res.Header = make(http.Header)
-	}
-
-	if res.Body == nil {
-		res.Body = &bodyBuffer{&bytes.Buffer{}}
-	}
-
-	c.servedWithResponse = true
-	c.res = res
-}
-
-// creates an empty shunt response with the initial status code of 404
-func shunt(r *http.Request) *http.Response {
-	return &http.Response{
-		StatusCode: http.StatusNotFound,
-		Header:     make(http.Header),
-		Body:       &bodyBuffer{&bytes.Buffer{}},
-		Request:    r}
-}
-
-// applies all filters to a request
-func (p *Proxy) applyFiltersToRequest(f []*routing.RouteFilter, ctx *filterContext, onErr func(err interface{})) []*routing.RouteFilter {
-	var start time.Time
 	var filters = make([]*routing.RouteFilter, 0, len(f))
 	for _, fi := range f {
-		start = time.Now()
-		tryCatch(func() { fi.Request(ctx) }, onErr)
-		p.metrics.MeasureFilterRequest(fi.Name, start)
+		start := time.Now()
+		tryCatch(func() {
+			fi.Request(ctx)
+			p.metrics.MeasureFilterRequest(fi.Name, start)
+		}, func(err interface{}) {
+			if p.flags.Debug() {
+				ctx.debugFilterPanics = append(ctx.debugFilterPanics, err)
+				return
+			}
+
+			log.Errorf("error while processing filter during request: %s: %v", fi.Name, err)
+		})
+
 		filters = append(filters, fi)
-		if ctx.served || ctx.servedWithResponse {
+		if ctx.deprecatedShunted() || ctx.shunted() {
 			break
 		}
 	}
+
+	p.metrics.MeasureAllFiltersRequest(ctx.route.Id, filtersStart)
 	return filters
 }
 
 // applies filters to a response in reverse order
-func (p *Proxy) applyFiltersToResponse(filters []*routing.RouteFilter, ctx filters.FilterContext, onErr func(err interface{})) {
+func (p *Proxy) applyFiltersToResponse(filters []*routing.RouteFilter, ctx *context) {
+	filtersStart := time.Now()
+
 	count := len(filters)
-	var start time.Time
 	for i, _ := range filters {
 		fi := filters[count-1-i]
-		start = time.Now()
-		tryCatch(func() { fi.Response(ctx) }, onErr)
-		p.metrics.MeasureFilterResponse(fi.Name, start)
+		start := time.Now()
+		tryCatch(func() {
+			fi.Response(ctx)
+			p.metrics.MeasureFilterResponse(fi.Name, start)
+		}, func(err interface{}) {
+			if p.flags.Debug() {
+				ctx.debugFilterPanics = append(ctx.debugFilterPanics, err)
```

Comment:
i'll add the docs, thx

to give an explanation, this was "practically" already like here, but instead of using a field on the context object, it was used as a variable. See the earlier filterPanics variable. The main reason of moving this variable to context object is that i am planning to open a pull request with a loopback backend. The context object is used only for a single request flow, and only evaluated by the debug scenario, so I believe it's just fine.